authentication design document. (#65)
chanzuckerberg/single-cell: rfcs/0001-authentication/text.md

   not part of this design.
1. The document describes authentication in both Data Portal and cellxgene and
   how they work together. The only feature that currently requires authentication
   in milestone 1 is the user annotation feature in cellxgene.
1. This document assumes that Auth0/OIDC/OAuth will be the technologies used
   for authentication, alternative authentication services are not considered
   here.
1. Cellxgene will continue to support the desktop stand alone version, the WSGI
   hosted version, and the CZI hosted version. Cellxgene must continue to work
   in all these modes. Nothing in this design breaks backwards compatibility or
   changes current user experience.

## Product Requirements

Cellxgene

1. [CXG1] The cellxgene client will provide an authenticated user with access
   to features that require authentication (just user annotations for now).
1. [CXG2] An authenticated user will have the option to logout.
1. [CXG3] An unauthenticated user will not have access to features that require
   authentication. They will be disabled in the app, and require login on first
   use.
1. [CXG4] An unauthenticated user will have an option to login

> 1. [CXG5] A user must not unexpectedly lose work (state of the client) when

Author: The requirements were meant to be neutral to the solution.   We had a few options here.
The key word is "unexpectedly".  If they have been warned then it's not unexpected...

   they login. Normally when a cellxgene dataset is accessed, there is a page
   reload, which presents the user with the initial state. Many authentication
   flows use a redirect, which would also reset the cellxgene state back to the
   initial state.
1. [CXG6] The cellxgene server will enforce the authentication for routes that
   require authentication by checking for the ID token, and rejecting requests
   that lack a token or have an invalid/expired token. The signature needs to be
   validated to ensure that the token has not been tampered with. An expired ID
   token can be refreshed (as long as the refresh token has not expired).

> 1. [CXG7] The cellxgene server will have an endpoint to check if the requestor

Author: The cellxgene frontend needs to know if the user is authenticated in order to show an Log In button or a Log Out button.  I can add that.  I have a new endpoint for this called "userinfo".

   is authenticated. The current (“config”) endpoint can serve this purpose by
   having an authentication section.
1. [CXG8] To avoid causing the ID token to expire while the user is working
   (and thus disrupting the user’s work), the token can be automatically refreshed
   when the ID token expires. The

>    expiration date of the Refresh token could be set to the maximum of 30 days.

Author: Actually, this should be indefinite now.  The refresh token is itself is refreshed when the id token expires.  So, a user can be logged in indefinitely as long as they visit the website at least once per 90 days.  I can add a comment for that.

   optionally configured for cellxgene
   deployments. If cellxgene is not configured for authentication then the login/logout buttons
   will not appear, and feature such as user-annotations will be controlled as they have been
   with configuration parameters (e.g. “user_annotations.enable”).
1. [CXG10] The cellxgene client code must not require a particular solution
   (i.e. Auth0). Auth0 code must not be added directly to the cellxgene client.
   Any such specialize code must be
   introduced using the modularity features of cellxgene. (In the design presented here there is
   no need to have such specializations in the cellxgene client).
1. [CXG11] The cellxgene server must not require any authentication specific
   modules to be installed if authentication is not configured. The cellxgene
   server may provide Auth0
   support as a configuration option, but anything specific to the deployment
   (client_id, client_secret, etc), must be configurable and not part of the code base.

Data Portal

1. [C1] An authenticated user will have the option to logout
1. [C2] An unauthenticated user will have an option to login

> 

Author: Yeah, portal and cellxgene work in the same way.   These should be reframed as common requirements.

Single sign on

1. [S1] A user authenticated in cellxgene will automatically be authenticated by Data Portal
1. [S2] A user authenticated in Data Portal will automatically be authenticated by cellxgene
1. [S3] Logging out of one app, logs the user out of both apps

General

> 1. [G1] GDPR requirements: For gdpr, the main thing we need to log is user id + time stamp for

Author: auth0 stores this.  We can get the logs or access to the their DB if we need to.

   when they first agreed to our terms of service. The underlying requirement is we need to log
   evidence of consent, ie the "click" of agreement to our tos.
1. [G2] The login experience requires that the user has consented to use cookies. It is assumed
   that this question has already been asked, and we first verify the answer before proceeding
   with the login process.

> ## Detailed Design | Architecture | Implementation

Reviewer: Sorry I forget, did we add any tables to the RDS for cellxgene to keep track of users? Or do we only store their identify for annotations?

automatically be logged into the other. Or they can log out of one, and automatically be
logged out of the other.

This will be achieved by using an httpOnly sameSite cookie, using the same domain for both
backends. Both applications will use the same Auth0 client id and domain. Auth0 will be
configured to register a callback to each of the two applications. A cookie setup this way
will be sent to the backend of both Data Portal and cellxgene during API requests. Both
applications will set this cookie on login, and remove it on logout.

For best security, both applications should use the same full domain name so that we may use the
“sameSite=strict” option on the cookie. For example: api.cellxgene.cziscience.com. This will
keep the two applications isolated. The API to each application can be distinguished using a
different directory within that domain:

Front-end:

1. cellxgene: cellxgene.cziscience.com/&lt;urlroot>/&lt;datasetname>/...
1. Data Portal: cellxgene.cziscience.com/

Back-end:

1. cellxgene: api.cellxgene.cziscience.com/cellxgene/...
1. Data Portal: api.cellxgene.cziscience.com/corpora/

> CloudFront (or any http proxy) can be configured to map the cellxgene API from the “api

Reviewer 2: I think it did

   includes the “Create new category” button, and the options to export or import user annotations.
1. When the user clicks the “Login” button the authentication flow starts.

When an authenticated user visits a dataset page:

1. A Logout button will be displayed at the top of the page
1. [Optional] we can tell the user who they are logged in as. This information can be made
   available in the configuration parameters (from the “config” endpoint).
1. The “Create new category” button is available, as well as UI elements to export and import
   annotations files, and any other UI elements that require authentication.
1. When the user clicks the “Logout” button, the logout flow starts.

Backward compatibility: If the annotation feature has been enabled, but authentication has not
been enabled, then the behavior is like the current cellxgene with annotations enabled. In this
case, there are no login/logout buttons. The user id is generated from the session id, and it
has no meaning outside of the current browser.

#### Login Flow

The flow is similar to “regular app” flow in the auth0 documentation. The “Web App” is the
cellxgene server in the diagram.

> ![Authentication Code Flow](imgs/authcodeflow.png)

Reviewer: One clarification question. In this diagram, does the "Browser" represent the cellxgene client browser or the browser that gets opened for Auth purposes?

> 

Reviewer: If you're willing/able to, I think a sequence UML diagram that showcases the steps for the login flow you've described below would be super helpful! It would show a clear picture of when the endpoints are called, by whom, and in what order. You could even substitute it for the above diagram which I (personally) don't find to be super helpful without cross checking with the wording below.

  with Cellxgene logos). The user can authenticate with an identity provider (like Google or others).

![Authentication Code Flow](imgs/authlogin.png)

- **[3]** Once the user has authenticated, Auth0 will direct the browser to the callback url
  in the cellxgene server. The callback url must be known to both Auth0 and to cellxgene. I
  suggest a name like “/oauth2/callback”. The request will contain the authorization code.
- **[4]** cellxgene can then exchange the authorization code for an ID token by sending a back
  channel request to Auth0. Cellxgene must know the client_secret for this step. Part of the
  configuration and startup process for cellxgene will be to retrieve this secret. In the case
  of CZI hosted cellxgene the client_secret will be retrieved from the AWS Secrets Manager.
- **[5]** Auth0 will send back the ID token, as a JWT (JSON Web Token).
- The cellxgene server then redirects (full page nav) the user back to the original dataset
  , and directs the client to save the JWT in an httpOnly cookie on the client.
- The JWT proves the identity of the client. Subsequent requests from the client to the

>   cellxgene server will include the cookie with the JWT. When the cellxgene client re-requests

Reviewer 2: ```suggestion
  cellxgene server will include the cookie with the `id_token`. When the cellxgene client re-requests
```
might be worth the extra specificity wrt the oauth2/oidc protocol

  the configuration through the dataset’s “config” endpoint (&lt;server>/d/&lt;datasetname>/api
  /v0.2/config), the configuration will now indicate that the user is authenticated. The
  cellxgene client can also have access to the user’s name, which it may use when displaying the log information.
  1. If the ID token is invalid, then the cellxgene server can re-initiate the authentication
     process by going back to step [1]
  2. The server will verify the signature of the token to ensure that it was signed by the
     sender and not altered.

>   3. If the ID token is expired, then the server will obtain a new ID token automatically

Reviewer 2: ```suggestion
  3. If the ID token is expired, then the client will obtain a new ID token automatically
```
I think this happens client-side?

- /login : this endpoint is used to initiate the login process. The client passes in the
  location to the current dataset in the request. The server then initiated the authentication
  request by redirecting the browser to the Auth0 universal login page. The server passes in the
  location of the dataset as a state variable in the request, so the browser can be redirected
  there after login. This is step 1 in the diagram. /login is requested when the user clicks on
  the “login” button.
- /logout : this endpoint is used to clear the authentication credentials. The cookie
  containing the ID token is removed. /logout is requested when the user clicks on the “logout” button.
- /oauth2/callback: this endpoint is accessed during step 3 of the diagram. Auth0 redirects
  the user to this endpoint, and the request contains the authentication code and the state
  parameter sent in from the /login endpoint. The cellxgene server then exchanges the
  authentication code for an ID Token. Once complete, the cellxgene server redirects the client
  back to the dataset, and directs the browser to store the ID Token in a cookie.

> 

Reviewer: I remember that the health/ API endoint ended up changing as well. Can you remark here (or maybe elsewhere is appropriate) why this change was needed and how it changed?

### Authentication in Data Portal

When a user visits the Data Portal site, we determine if the user is authenticated or not by

> querying a backend endpoint (e.g. api.cellxgene.cziscience.com/corpora/is-authenticated). This

Reviewer: Was this built?

Author: It's called "userinfo" in the implementation.   Yes, it's built.

can return back the username if authenticated, or a null response if not authenticated.

If authenticated, a Logout button may appear near the top of the page.

If unauthenticated, a Login button may appear near the top of the page, with the name of the user.

In a previous version of this design doc, the authentication strategy for Data Portal was to use
the [Authentication Code Flow with Proof Key for Code Exchange (PKCE)](https://auth0.com/docs/flows/concepts/auth-code-pkce),
implemented in the client side. This has now been changed to
using the Authentication Code flow in the backend (same as cellxgene). This approach is
simpler with respects to the single sign on strategy. It avoids a step where the client

> needs to access the server to set the cookie containing the tokens. It is also more secure.

Reviewer: Why is it more secure?

The Data Portal front end requires backend support for the following endpoints

- Login: start the authentication flow, set the cookie when complete
- Logout: logout, remove the cookie

> - Is-authenticated: check if authenticated, and return user information

Reviewer: Sorry, I can't remember if this was built or not, but can you replace with the actual endpoint name if it was?

Option 2 is the best choice. The amount of coding is small enough when using libraries like authlib.

### Future Work

The following lists possible work that can be done after the design described above has been implemented.

1. if the user has declined cookies, we could (with some work), still enable a temporary login
   while the user remains on the current page. Single sign one would not work in this case. I’m
   not advocating for this option.

## Alternatives

### Cellxgene could use a client side approach such as authorization code with proof key code exchange (PKCE)

The login/authentication approach described in the cellxgene section uses the more secure
Authorization Code grant type (response_type=code). This flow requires extra steps (steps [4]
and [5] in the diagram), where the authorization code is exchanged to an ID token by making a
request from the cellxgene server to the Auth0 server.

An alternative is to use the authorization code with proof key code exchange (PKCE). This is
the approach you might use in a single page app where you do not want the code running in a user’s

> browser to have access to the auth0 client secret.

Reviewer 2: ```suggestion
browser to have access to the oauth2 client secret.
```

“The implicit grant (response type "token") and other response types causing the authorization
server to issue access tokens in the authorization response are vulnerable to access token
leakage and access token replay as described in [Section 3.1](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-3.1), [Section 3.2](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-3.2), [Section 3.3](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-3.3), and [Section 3.6](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-09#section-3.6). In order to avoid these issues, Clients SHOULD NOT use
the implicit grant and any other response type causing the authorization server to issue an
access token in the authorization response.”

The popup experience may also depend on the client’s browser and security settings, which could
lead a larger testing and support burden.

So, the consensus was to go with the “regular app” approach, and strongly encourage the user to
login before doing work.

## References

> [0][location to the authorization code flow diagram](https://auth0.com/docs/architecture-scenarios/web-app-sso/part-1)

Reviewer 3: @bmccandless if we want the link text not be lowercase, we'll need to add `\` before the number like below, otherwise linter will force lowercase! Thank you!

```md
## References
\[0][IETF's Github Process](https://tools.ietf.org/html/draft-ietf-git-using-github-06)
\[1][Google Developer Documentation Style Guide](https://developers.google.com/style/highlights#introduction_1)
\[2][Data Coordination Platform RFC Process](https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0001-rfc-process.md)
```
